feat: add property to disable FunctionProperty caching

vlsi opened the pull request:
## Description

By default, JMeter caches function properties during a test iteration, however, it might cause unexpected results when a component is shared across threads and the expression depends on the thread variables.

It looks like the caching results in wrong HTTP headers being sent if https://github.com/apache/jmeter/pull/727 is applied.
Even though JMeter does not modify HTTP Header Manager components during the execution, however, it looks like any function that is used in the headers would be "cached", so other threads could reuse the value generated by a completely different thread.

I believe the current caching behaviour is not well-defined, so I suggest we disable it in the upcoming versions.

The PR does not change the existing behaviour, and it just adds a property so the users could disable caching.

vlsi committed:
fix: disable FunctionProperty caching

Previously it cached the values based on iteration number only
which triggered wrong results on concurrent executions.

The previous behavior can be temporary restored with
function.cache.per.iteration=true property.

FSchumacher commented on xdocs/usermanual/properties_reference.xml:
```diff
@@ -1339,6 +1339,14 @@ JMETER-SERVER</source>
+    <p>By default, JMeter caches function properties during a test iteration, however,
+    it might cause unexpected results when a component is shared across threads and the expression depends on
+    the thread variables.</p>
+    <note>The caching behaviour would likely be disabled in the upcoming versions</note>
```
```suggestion
    <note>The caching behaviour will likely be disabled in an upcoming version</note>
```
Looks good to me.
Out of curiosity, have you done performance comparisons? I would think, that in most cases on modern CPUs the cache is not that important anymore.

vlsi replied:
I believe the cache creates functional issues: it caches the values, and it does not invalidate the cache. In other words, it assumes "the computed value depends only on the iteration number", which, in my opinion, is quite poor approximation.

So my main goal here is to establish correctness, and if there are any performance implications, then they should be healed separately

vlsi replied:
I have not done explicit comparisons, however, I saw no perf difference when it comes to my regular load testing results

vlsi commented:
Frankly speaking, I incline we should change the default to "do not cache based on iteration number only" as the previous behavior (==cache based on iteration number) was introducing silent and hard-to-notice bugs, so I would alter the default to "do not cache".
If somebody still needs the cache, they could activate it back, however, I expect we would still remove the setting sometime in the future.

vlsi committed:
fix: avoid calling getScript and getFilename two times in BeanShellSampler, and JSR223TestSampler

Previously, BeanShellSampler called getScript several times which
might cause wrong results when the script included a function that
changes on each call (e.g. couter).
Now getScript is called only once per sample.

vlsi commented on src/protocol/java/src/main/java/org/apache/jmeter/protocol/java/sampler/BeanShellSampler.java:
```diff
@@ -102,14 +102,6 @@ public SampleResult sample(Entry e)// Entry tends to be ignored ...
             return res;
         }
         try {
-            String request = getScript();
```
@FSchumacher , it turns out this caused two executions of `getScript` for `BeanShellSampler`.
One was to pass the script for `res.setSamplerData`, and the second `getScript` was in `processFileOrScript` for actual BSH execution.

I guess that is unexpected. However, the fix does not seem to be trivial: if I change `processFileOrScript` signature, it breaks backward compatibility because others might have overridden the method.
On the other hand, `processFileOrScript` does not expose `script` and `fileName`, so it is not clear how to get that information out.

For now, I added a new parameter to `processFileOrScript` so it calls `setSamplerData` as needed. I think it is unlikely that people override `processFileOrScript`. I checked GitHub search, and I found no cases for overriding `processFileOrScript`.

If you are ok with that, I think we need to make the same changes for JSR223 samplers. They have exactly the same issue, and it can call `getScript()` several times during execution as well.
